doc/data-modules/defining-data-types.rst

```diff
@@ -0,0 +1,84 @@
+Defining data types
+===================
+
+Data types can be defined using the ``declareType()`` method. It expects a name
+(which you can later use with ``storeObject()``), as well as a `JSON Schema`_
+object defining the actual structure and formatting of your data.
+
+Consider this simplified example of an archive bookmark:
+
+.. CODE:: javascript
+
+   var Bookmarks = { name: 'bookmarks', builder: function(privateClient, publicClient) {
+
+     privateClient.declareType('archive-bookmark', {
+       "type": "object",
+       "properties": {
+         "id": {
+           "type": "string"
+         },
+         "title": {
+           "type": "string"
+         },
+         "url": {
+           "type": "string",
+           "format": "uri"
+         },
+         "tags": {
+           "type": "array",
+           "default": []
+         },
+       },
+       "required": [ "title", "url" ]
+     });
+
+     // ...
+   }};
+
+Now that we have a basic data type in place for storing bookmarks, we can add a
+function for storing them. This will actually validate the incoming data
+against the type's schema, and reject the promise with detailed validation
+errors in case the data format doesn't match::
+
+   var Bookmarks = { name: 'bookmarks', builder: function(privateClient, publicClient) {
+     // ...
+
+     return {
+       exports: {
+
+         add: function (bookmark) {
+           bookmark.id = md5Hash(bookmark.url); // hash URL for nice ID
```

Comment:
This is the only part of the example that will not work out of the box by copying and pasting (since the `md5Hash` function isn't defined). I think including a short hashing function or saying in the comment that this is not expected to be working code would be better. What do you think?

Reply:
I never thought anyone would try this as working code. In fact, that's a bad idea, as it would store data right next to actual Webmarks data, which has different requirements.

I don't know what to do about it, but it's actually good that this is failing as is.

Reply:
Good point, I only tried it as part of the review to check the syntax. Since it's not even a complete snippet I actually think it's good like this

Reply:
Yer, the next thing that would fail is `storeObject`, because the type is not defined in the same snippet.